src/sbvr-api/control-flow.ts

```diff
@@ -0,0 +1,61 @@
+import * as _ from 'lodash'
+import * as Promise from 'bluebird'
+
+// lifts f to map over the changeSet if present
+export const liftP = <T, U>(fn: (v: T) => U | Promise<U>) => {
+	return ((a: T | T[]) => {
+		if (_.isArray(a)) {
+			// This must not be a settle as if any operation fails in a cs
+			// we want to discard the whole
+			return Promise.mapSeries(a, fn)
+		} else {
+			return Promise.resolve(a).then(fn)
+		}
+	// TODO: We cast here to merge `Promise<U> | Promise<U[]>` into `Promise<U | U[]>,
+	// which is equivalent for our uses but fixes issues where typing was being rejected
+	}) as ((a: T | T[]) => Promise<U | U[]>)
+}
+
+interface MappingFunction {
+	<T, U>(a: T[], fn: (v: T) => U | U[] | Promise<U>): Promise<Array<U | U[] | Error>>
+}
```

Comment:
I think you can drop the `U[]` in here. 
```
interface MappingFunction {
	<T, U>(a: T[], fn: (v: T) => U | Promise<U>): Promise<Array<U | Error>>
}
```
Works for me

Reply:
I think I added it because of sbvr-utils usage needing it, I'll check that

Reply:
It isn't immediately obvious the the `U[]` is required so I've removed it and we can cross that bridge later (also I updated bluebird definitions and the `U[]` started causing issues.. so it definitely needs to be done differently if it does need to be done)